### DISCUSSION TITLE
Question about the "index" parameter of the add() method on BeliefBase implementations?

### DISCUSSION THREAD

**Discussion by mindcrime** [jason-lang/jason#120] [Q&A] [answered]
Hi all, apologies for the probably terribly simple and obvious question, but could anybody tell me more about what the "index" parameter to the BeliefBase add() method is for?

I see we have two overrides of add(), like

`
public boolean add( int index, Literal l )
`
and
`
public boolean add(Literal l )
`

and it's not immediately clear to me what the index is used for. Is it as simple as allowing the same belief to appear multiple times (and how would that work) or is there something more interesting going on here?

EDIT:

OK, reading [the javadoc](https://jason-lang.github.io/api/jason/bb/BeliefBase.html#add(int,jason.asSyntax.Literal)) I see this bit:

```
public boolean add(int index,
 [Literal](https://jason-lang.github.io/api/jason/asSyntax/Literal.html) l)
            throws [JasonException](https://jason-lang.github.io/api/jason/JasonException.html)
Adds a belief in the BB at index position, returns true if succeed
```

for the version that takes the index parameter. As opposed to the other override, which says:

```
public boolean add([Literal](https://jason-lang.github.io/api/jason/asSyntax/Literal.html) l)
            throws [JasonException](https://jason-lang.github.io/api/jason/JasonException.html)
Adds a belief in the end of the BB, returns true if succeed. <snip>
```

So it seems like the only difference is, if you call the single parameter version of the add() function, the new belief is added at the end of the BB by default, and if you pass an index, you get to specify the position in the BB? If I'm reading that right, I guess my follow-on question would be "when/why would I want to specify the index explicitly?"  And, "is this useful for customized BB implementations that don't have an obvious notion of "position" like a JDBC backed BB, or one using an RDF triplestore? 

Thanks in advance!

**Comment by jomifred** [comment, marked as answer]
Hi,

first, Jason default implementation of BB only support add in the beginning or in the ending of the BB (other implementation may implement other uses of the `index`). 

The order is quite important for the queries. For instance, suppose a BB with

```
b(10).
b(20).
```

a query `.findall(X,b(X),L)` should return `L=[10,20]`, with this order of values. 

Moreover, if the agent runs `+b(30)` followed by (maybe in another intention) `?b(X)`, I'd expect that `X=30`, the most recent value. Well, at least, we thought that is more "natural" :-) The `+` operator adds the belief in the beginning of the BB. Anyway, the operator `+>` can be used to some belief in the end of the BB (see [operators](https://github.com/jason-lang/jason/blob/main/doc/tech/Jason%20Operators%20and%20Events.pdf)).

I don't know if that reasons also apply to RDF....

**Comment by mindcrime** [comment]
OK, I think I get it. Sounds like anyone implementing a custom BeliefBase should then store the index parameter (when provided) and then when returning query results sort by that index. I assume the index is intended to start at 0 (or 1) and increase monotonically? 

One other further question if y'all will bear with me here: is the index per-belief, or global across all beliefs in the BB?

Thanks!

**Comment by jomifred** [comment]
Index starts at 0 and is related to a belief type (not all BB). belief type is functor+arity (class PredicateIndicator).

**Reply by mindcrime** [reply]
Rad. Thanks again!
